client/network/src/gossip.rs

```diff
@@ -180,66 +173,52 @@ impl<'a, M: Send + 'static> QueueGuard<'a, M> {
 
 impl<'a, M> Drop for QueueGuard<'a, M> {
 	fn drop(&mut self) {
-		// We notify the `Condvar` in the destructor in order to be able to push multiple
-		// messages and wake up the background task only once afterwards.
-		self.condvar.notify_one();
+		let _ = self.notify_background_future.try_send(());
 	}
 }
 
 #[derive(Debug)]
 struct Shared<M> {
-	/// Read by the background task after locking `locked`. If true, the task stops.
-	stop_task: atomic::AtomicBool,
 	/// Queue of messages waiting to be sent out.
 	messages_queue: Mutex<VecDeque<M>>,
-	/// Must be notified every time the content of `locked` changes.
-	condvar: Condvar,
 	/// Maximum number of elements in `messages_queue`.
 	queue_size_limit: usize,
 }
 
 async fn spawn_task<B: BlockT, H: ExHashT, M, F: Fn(M) -> Vec<u8>>(
+	mut wait_for_sender: Receiver<()>,
 	service: Arc<NetworkService<B, H>>,
 	peer_id: PeerId,
 	protocol: ConsensusEngineId,
 	shared: Arc<Shared<M>>,
 	messages_encode: F,
 ) {
 	loop {
-		let next_message = 'next_msg: loop {
-			let mut queue = shared.messages_queue.lock().await;
-
-			loop {
-				if shared.stop_task.load(atomic::Ordering::Acquire) {
-					return;
-				}
-
-				if let Some(msg) = queue.pop_front() {
-					break 'next_msg msg;
-				}
-
-				// It is possible that the destructor of `QueuedSender` sets `stop_task` to
-				// true and notifies the `Condvar` after the background task loads `stop_task`
-				// and before it calls `Condvar::wait`.
-				// See also the corresponding comment in `QueuedSender::drop`.
-				// For this reason, we use `wait_timeout`. In the worst case scenario,
-				// `stop_task` will always be checked again after the timeout is reached.
-				queue = shared.condvar.wait_timeout(queue, Duration::from_secs(10)).await.0;
-			}
-		};
-
-		// Starting from below, we try to send the message. If an error happens when sending,
-		// the only sane option we have is to silently discard the message.
-		let sender = match service.notification_sender(peer_id.clone(), protocol) {
-			Ok(s) => s,
-			Err(_) => continue,
-		};
-
-		let ready = match sender.ready().await {
-			Ok(r) => r,
-			Err(_) => continue,
-		};
-
-		let _ = ready.send(messages_encode(next_message));
+		if wait_for_sender.next().await.is_none() {
```

Comment:
I don't think that works. You're blocking indefinitely here.

Reply:
Can you add more details why that is a problem @tomaka? Reaching this line implies that there are currently no more messages to send. Thus blocking until the sender adds more messages should be the right approach.

Reply:
I see, I totally misread the code above. Nevermind.